action/governance/withdrawFunds.go

```diff
@@ -0,0 +1,234 @@
+	if err != nil || !ok {
+		ctx.Logger.Error("Failed to delete proposal from ACTIVE store :", proposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "failed_to_delete_proposal_from_active_store"),
+		}
+		return false, result
+	}
+
+	// 3. Check if the contributor has funded this proposal, if so, get the amount of funds
+	proposalFund, err := governance.GetCurrentFundsByContributor(proposal.ProposalID, withdrawProposal.Contributor, ctx.ProposalMasterStore.ProposalFund)
+	if err != nil {
+		ctx.Logger.Error("No available funds to withdraw for this contributor :", withdrawProposal.Contributor)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "no_available__fund_to_withdraw_for_this_contributor"),
+		}
+		return false, result
+	}
+
+	// 4. Check if the contributor has sufficient funds to withdraw for that proposal
+	withdrawAmount := balance.NewAmountFromBigInt(withdrawProposal.WithdrawValue.Value.BigInt())
+	if proposalFund.BigInt().Cmp(withdrawAmount.BigInt()) < 0 {
+		ctx.Logger.Error("Insufficient funds to withdraw for this contributor :", withdrawProposal.Contributor)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "no_available__fund_to_withdraw_for_this_contributor"),
+			Log: action.ErrNotEnoughFund.Msg,
+		}
+		return false, result
+	}
+
+	// 5. withdraw
+	// deduct from proposal fund
+	err = ctx.ProposalMasterStore.ProposalFund.DeductFunds(proposal.ProposalID, withdrawProposal.Contributor, withdrawAmount)
+	if err != nil {
+		ctx.Logger.Error("Failed to deduct funds from proposal:", withdrawProposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_deduct_fund_failed"),
+		}
+		return false, result
+	}
+	// add to beneficiary address
+	coin := withdrawProposal.WithdrawValue.ToCoin(ctx.Currencies)
+	err = ctx.Balances.AddToAddress(withdrawProposal.Beneficiary.Bytes(), coin)
+	if err != nil {
+		// return funds to proposal
+		err = ctx.ProposalMasterStore.ProposalFund.AddFunds(proposal.ProposalID, withdrawProposal.Contributor, withdrawAmount)
+		if err != nil {
+			ctx.Logger.Error("Failed to return funds to proposal:", withdrawProposal.ProposalID)
+			result := action.Response{
+				Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_return_fund_failed"),
+			}
+			return false, result
+		}
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_addition_failed"),
+		}
+		return false, result
+	}
+
+	result := action.Response{
+		Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_success"),
+	}
+
+	// check proposal funds:
+	// if all funds for this proposal has been withdrawn, mark outcome cancelled
+	if currentFundsForProposal.BigInt().Cmp(big.NewInt(0)) == 0 {
+		proposal.Outcome = governance.ProposalOutcomeCancelled
```

Comment:
ProposalOutcomeInsufficientFunds in this situation?

Reply:
We have a proposal cancel Tx out there.

Reply:
And we won't get proposal updated without writing to the store, right?

Reply:
sorry this one should be deleted, it was my previous design